core/src/avm2/verify.rs

```diff
@@ -517,6 +521,109 @@ fn verify_code_starting_from<'gc>(
     Ok(())
 }
 
+#[derive(Clone, Copy, Debug)]
+enum ValueType<'gc> {
+    // Either a class, or null.
+    Class(ClassObject<'gc>),
+    Int,
+    Uint,
+    Number,
+    Boolean,
+    Null,
+    Any,
+}
+
+#[derive(Clone, Debug)]
+struct Locals<'gc>(Vec<ValueType<'gc>>);
+
+impl<'gc> Locals<'gc> {
+    fn new(size: usize) -> Self {
+        Self(vec![ValueType::Any; size])
+    }
+
+    fn set_class_object(&mut self, index: usize, class: ClassObject<'gc>) {
+        self.0[index] = ValueType::Class(class);
+    }
+
+    fn set_any(&mut self, index: usize) {
+        self.0[index] = ValueType::Any;
+    }
+
+    fn set(&mut self, index: usize, value: ValueType<'gc>) {
+        self.0[index] = value;
+    }
+
+    fn at(&self, index: usize) -> Option<ValueType<'gc>> {
```

Comment:
Since the earlier verifier step checked the index, I think it should be either non-Option, or - if we intend to merge the abstract interpreter with the verifier (as the verifier also checks for type merging and stack state, which come from the abstract interpreter) - directly throw a VerifyError from here?

Reply:
Since we aren't merging verifier and optimizer yet, and the optimizer goes through bytecode unaware of control flow layout, it can hit unreachable code that the verifier ignored